src/ledger/database/lmdb.zig

```diff
@@ -0,0 +1,716 @@
+                        start_operation = .prev;
+                    },
+                }
+            } else if (direction == .reverse) {
+                _ = try cursorGetRelative(cursor, .last);
+            }
+
+            return .{
+                .allocator = self.allocator,
+                .txn = txn,
+                .cursor = cursor,
+                .direction = switch (direction) {
+                    .forward => .next,
+                    .reverse => .prev,
+                },
+                .next_operation = start_operation,
+            };
+        }
+
+        pub fn Iterator(cf: ColumnFamily, _: IteratorDirection) type {
+            return struct {
+                allocator: Allocator,
+                txn: *c.MDB_txn,
+                cursor: *c.MDB_cursor,
+                direction: CursorRelativeOperation,
+                next_operation: CursorRelativeOperation,
+
+                /// Calling this will free all slices returned by the iterator
+                pub fn deinit(self: *@This()) void {
+                    c.mdb_cursor_close(self.cursor);
+                    c.mdb_txn_abort(self.txn);
+                }
+
+                pub fn next(self: *@This()) anyerror!?cf.Entry() {
+                    const key, const val = try self.nextImpl() orelse return null;
+                    return .{
+                        try key_serializer.deserialize(cf.Key, self.allocator, key),
+                        try value_serializer.deserialize(cf.Value, self.allocator, val),
+                    };
+                }
+
+                pub fn nextKey(self: *@This()) anyerror!?cf.Key {
+                    const key, _ = try self.nextImpl() orelse return null;
+                    return try key_serializer.deserialize(cf.Key, self.allocator, key);
+                }
+
+                pub fn nextValue(self: *@This()) anyerror!?cf.Value {
+                    _, const val = try self.nextImpl() orelse return null;
+                    return try key_serializer.deserialize(cf.Value, self.allocator, val);
+                }
+
+                /// Returned data does not outlive the iterator.
+                pub fn nextBytes(self: *@This()) LmdbError!?[2]BytesRef {
+                    const key, const val = try self.nextImpl() orelse return null;
+                    return .{
+                        .{ .allocator = null, .data = key },
+                        .{ .allocator = null, .data = val },
+                    };
+                }
+
+                fn nextImpl(self: *@This()) LmdbError!?struct { []const u8, []const u8 } {
+                    defer self.next_operation = self.direction;
+                    return try cursorGetRelative(self.cursor, self.next_operation);
+                }
+            };
+        }
+    };
+}
+
+fn toVal(bytes: []const u8) c.MDB_val {
+    return .{
+        .mv_size = bytes.len,
+        .mv_data = @constCast(@ptrCast(bytes.ptr)),
```

Comment:
I dunno about this. While I see what you're doing with this constCast, it can lead to very illegal behavior. One of the *only* use cases of constCast is to mitigate incorrect APIs, which I actually don't think this is. 
I bet the implementation stores a mutable pointer because it frees it at some point. I have not looked into lmdb, so don't quote me on that.

Please take a `[]u8` here and duplicate and constant strings or string literals instead. I refuse to use constCast anywhere in this codebase.

Reply:
I'm resolving this comment since it is a duplicate